File: DotNet/Tests/Mocks/MockStore.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Reactive.Concurrency;
using System.Web;
using LoveSeat.Interfaces;
using Newtonsoft.Json.Linq;
using Shastra.Hydra.Messaging.MessageIds;
using Shastra.Hydra.Messaging.Storage;

namespace Shastra.Hydra.Tests.Mocks
{
    internal class MockStore : IStore
    {
        private readonly string _suffix;
        private readonly List<DocInfo> _docInfos = new List<DocInfo>();
        private readonly Dictionary<string, JToken> _docs = new Dictionary<string, JToken>();
        private readonly IScheduler _scheduler;
        private readonly object _lock = new object();
        private long _lastMicroseconds = 0;

        public string Name { get; private set; }

        public MockStore(string name, string suffix = null, IScheduler scheduler = null)
        {
            Name = name;
            _suffix = suffix ?? "";
            _scheduler = scheduler ?? Scheduler.CurrentThread;
        }

        public IEnumerable<IMessageId> GetChanges(IMessageId startId, long sinceSeq, out long lastSeq)
        {
            // Get changes after sinceSeq, and drop messages at or before _startId
            Replicate();
            lastSeq = GetLastSeq();
            return Enumerable.Range((int) sinceSeq + 1, (int) (lastSeq - sinceSeq)).Select(seqId => MessageIdManager.Create(_docInfos[seqId].DocId)).
                OrderBy(mId => mId).SkipWhile(mId => mId.CompareTo(startId) <= 0);
        }

        public long GetLastSeq()
        {
            // SeqId is the index of the last element of _docs
            return _docs.Count - 1;
        }

        public IMessageId SaveDoc(string json)
        {
            Replicate();

            JToken doc = JToken.Parse(json);
            DateTime idDate;
            string docId;
            // Allow the default DocId to be overridden in a TestHydraMessage
            if (doc["idDate"] != null) {
                idDate = DateTime.Parse((string) doc["idDate"]);
                // This bypasses the uniqueness check, so don't specify the same date twice.
                docId = ((long) (idDate.Subtract(new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)).TotalMilliseconds) * 1000).ToString("x14") + _suffix;
            } else {
                idDate = _scheduler.Now.UtcDateTime;
                docId = CreateDocId(idDate);
            }
            JToken stored = new JObject(new JProperty("id", docId), new JProperty("doc", doc));
            var docInfo = new DocInfo(docId, (string) doc["topic"], (string) doc["destination"], idDate);
            // Lock out other SaveDoc calls so that we definitely get the right list length
            lock (_lock) {
                stored["value"] = _docInfos.Count;
                _docInfos.Add(docInfo);
            }
            _docs[docId] = stored;
            return MessageIdManager.Create(docId);
        }

        public ServerDistanceInfo MeasureDistance()
        {
            return new ServerDistanceInfo { Name = Name, Distance = 10, IsReachable = true };
        }

        public IEnumerable<JToken> GetDocs(string viewName, IViewOptions options)
        {
            // We are only interested in viewNames "broadcastMessages" and "directedMessages"
            // options always has IncludeDocs=true and either:
            //   Keys = list of keys to get, each all [topic, couchId] or all [topic, destination, couchId] respectively. Or
            //   StartKey = [topic, couchId] or [topic, destination, couchId] and EndKey = [topic, maxvalue] or [topic, destination, maxvalue] respectively

            // Extract topic and possibly destination from here
            JArray filterArray;
            string startId = null;
            HashSet<string> keySet = null;
            if (options.StartKey.HasValues) {
                filterArray = JArray.Parse(HttpUtility.UrlDecode(options.StartKey.ToString()));
                startId = (string) filterArray.Last;
            } else {
                var keyArray = options.Keys.Select(key => JArray.Parse(HttpUtility.UrlDecode(key.ToString()))).ToList();
                // Empty array of keys for some reason
                if (!options.Keys.Any()) return Enumerable.Empty<JToken>();
                filterArray = keyArray[0];
                keySet = new HashSet<string>(keyArray.Select(key => (string) key.Last));
            }
            string topic = (string) filterArray[0];

            // Find the docs matching topic and possibly destination
            IEnumerable<DocInfo> filteredDocInfos;
            if (viewName == "broadcastMessages") {
                filteredDocInfos = _docInfos.Where(di => di.Topic == topic && di.Destination == null);
            } else if (viewName == "directedMessages") {
                string destination = (string) filterArray[1];
                filteredDocInfos = _docInfos.Where(di => di.Topic == topic && di.Destination == destination);
            } else {
                throw new Exception(string.Format("MockStore.GetDocs: View name {0} is not implemented.", viewName));
            }

            // Return either those on or after startId, or in keySet
            if (startId != null) {
                filteredDocInfos = filteredDocInfos.Where(di => String.Compare(di.DocId, startId, StringComparison.InvariantCulture) >= 0);
            } else {
                filteredDocInfos = filteredDocInfos.Where(di => keySet.Contains(di.DocId));
            }
            return filteredDocInfos.OrderBy(di => di.DocId).Select(di => _docs[di.DocId]);
        }

        private void Replicate()
        {
            // For now do nothing
        }

        /// <summary>
        /// Create a docid based on the date, incorporating the suffix, and ensuring monotonicity and uniqueness.
        /// </summary>
        /// <param name="date">The UTC date</param>
        /// <returns>A unique docid, higher in the sort order than any previous one</returns>
        private string CreateDocId(DateTime date)
        {
            long microseconds = (long) (date.Subtract(new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc)).TotalMilliseconds) * 1000;
            _lastMicroseconds = (microseconds > _lastMicroseconds) ? microseconds : _lastMicroseconds + 1;
            return _lastMicroseconds.ToString("x14") + _suffix;
        }

    }
}

File: Java/hydra/src/test/java/uk/co/shastra/hydra/messaging/mocks/MockStore.java
package uk.co.shastra.hydra.messaging.mocks;

import java.text.ParsePosition;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.TreeSet;

import org.ektorp.ComplexKey;
import org.ektorp.ViewQuery;

import rx.Scheduler;
import rx.concurrency.CurrentThreadScheduler;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeType;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.POJONode;

import uk.co.shastra.hydra.messaging.messageids.MessageId;
import uk.co.shastra.hydra.messaging.messageids.MessageIdManager;
import uk.co.shastra.hydra.messaging.storage.ChangesWrapper;
import uk.co.shastra.hydra.messaging.storage.ServerDistanceInfo;
import uk.co.shastra.hydra.messaging.storage.Store;

public class MockStore implements Store {

	private String name;
	private String suffix;
	private Scheduler scheduler;
    private ArrayList<DocInfo> docInfos = new ArrayList<DocInfo>();
    private HashMap<String, JsonNode> docs = new HashMap<String, JsonNode>();
	private long lastMicroseconds = 0;

	@Override public String getName() { return name; }

    public MockStore(String name) { this(name, null, null);}
    public MockStore(String name, String suffix) { this(name, suffix, null); }
    public MockStore(String name, String suffix, Scheduler scheduler)
    {
        this.name = name;
        this.suffix = suffix == null ? "" : suffix;
        this.scheduler = scheduler == null ? CurrentThreadScheduler.getInstance() : scheduler;
    }
    
	@Override
	public ChangesWrapper getChanges(MessageId startId, long sinceSeq) {
        // Get changes after sinceSeq, and drop messages at or before _startId
        Replicate();

        long lastSeq = getLastSeq();
        TreeSet<MessageId> idSet = new TreeSet<MessageId>();
        for (long ii = sinceSeq + 1; ii < lastSeq + 1; ii ++) {
        	MessageId messageId = MessageIdManager.create(docInfos.get((int) ii).docId);
        	if (messageId.compareTo(startId) > 0) idSet.add(messageId);
        }
		return new ChangesWrapper(idSet, lastSeq);
	}

	@Override
	public long getLastSeq() {
        // SeqId is the index of the last element of _docs
        return docs.size() - 1;
	}

	@Override
	public MessageId saveDoc(JsonNode json) {
        Replicate();

        Date idDate;
        String docId;
        // Allow the default DocId to be overridden in a TestHydraMessage
        if (json.get("idDate") != null) {
            idDate = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ").parse(json.get("idDate").textValue(), new ParsePosition(0));
            // This bypasses the uniqueness check, so don't specify the same date twice.
            docId = String.format("%1$014x%2$s", idDate.getTime() * 1000, suffix);
        } else {
            idDate = new Date(scheduler.now());
            docId = CreateDocId(idDate);
        }
        // The following is not good. It's based on the .NET MockStore, where matching the behaviour of LoveSeat means we want to store and return
        // a JSON object containing _id, value and doc, where doc is the document being saved. However Ektorp returns a document element containing
        // the _id and value fields as fields of the doc. So I've kept the .NET model, but bodged the extra fields into doc. This could be simplified.
        ((ObjectNode) json).put("_id", docId);
        ObjectNode stored = new ObjectMapper().createObjectNode();
        stored.put("doc", json);
        stored.put("_id", docId);
        String topic = json.has("topic") ? json.get("topic").asText() : null;
        String destination = json.has("destination") ? json.get("destination").asText() : null;
        DocInfo docInfo = new DocInfo(docId, topic, destination, idDate);
        // Lock out other SaveDoc calls so that we definitely get the right list length
        synchronized (this) {
            stored.put("value", docInfos.size());
            ((ObjectNode) json).put("value", docInfos.size());
            docInfos.add(docInfo);
        }
        docs.put(docId, stored);
        return MessageIdManager.create(docId);
	}

	@Override
	public Iterable<JsonNode> GetDocs(String viewName, ViewQuery options) {
        // We are only interested in viewNames "broadcastMessages" and "directedMessages"
        // options always has IncludeDocs=true and either:
        //   Keys = list of keys to get, each all [topic, couchId] or all [topic, destination, couchId] respectively. Or
        //   StartKey = [topic, couchId] or [topic, destination, couchId] and EndKey = [topic, maxvalue] or [topic, destination, maxvalue] respectively

        // Extract topic and possibly destination from here
        ArrayNode filterArray;
        String startId = null;
        HashSet<String> keySet = null;
        if (options.getStartKey() != null) {
            filterArray = (ArrayNode) ((ComplexKey) options.getStartKey()).toJson();
            startId = JsonNodeToString(filterArray.get(filterArray.size() - 1));
        } else {
        	ArrayNode keyArray = null;
			try {
				// getKeysAsJson gives a String like "{\"keys\" : [key0, key1, ...] }" where keyi is of the form "[val0, val1, ...]"
				keyArray = (ArrayNode) (new ObjectMapper().readTree(options.getKeysAsJson()).get("keys"));
			} catch (Exception e) {}
            // Empty array of keys for some reason
            if (keyArray.size() == 0) return new ArrayList<JsonNode>();
            filterArray = (ArrayNode) keyArray.get(0);
            keySet = new HashSet<String>(); //keyArray.Select(key => (string) key.Last));
            for (JsonNode key : keyArray) {
            	ArrayNode keyAsArray = (ArrayNode) key;
            	keySet.add(JsonNodeToString(keyAsArray.get(keyAsArray.size() - 1)));
            }
        }
        String topic = JsonNodeToString(filterArray.get(0));

        // Find the docs matching topic and possibly destination
        ArrayList<DocInfo> filteredDocInfos = new ArrayList<DocInfo>();
        if (viewName.equals("broadcastMessages")) {
        	for (DocInfo di : docInfos) {
				if (di.topic.equals(topic) && di.destination == null) filteredDocInfos.add(di);
			}
        } else if (viewName.equals("directedMessages")) {
            String destination = JsonNodeToString(filterArray.get(1));
        	for (DocInfo di : docInfos) {
				if (di.topic.equals(topic) && di.destination == destination) filteredDocInfos.add(di);
			}
        } else {
            // This view isn't implemented
        }

        // Return either those on or after startId, or in keySet
        TreeSet<String> filteredDocIds = new TreeSet<String>();
        if (startId != null) {
        	for (DocInfo di : filteredDocInfos) {
				if (di.docId.compareTo(startId) >= 0) filteredDocIds.add(di.docId); 
			}
        } else {
        	for (DocInfo di : filteredDocInfos) {
				if (keySet.contains(di.docId)) filteredDocIds.add(di.docId); 
			}
        }
        ArrayList<JsonNode> res = new ArrayList<JsonNode>();
        for (String docId : filteredDocIds) {
			res.add(docs.get(docId).get("doc"));
		}
        return res;
	}

	@Override
	public ServerDistanceInfo measureDistance() {
		return new ServerDistanceInfo(getName(), true, 10);
	}

    private void Replicate()
    {
        // For now do nothing
    }
    
    /**
     * Create a docid based on the date, incorporating the suffix, and ensuring monotonicity and uniqueness.
     * 
     * @param date The UTC date
     * @return A unique docid, higher in the sort order than any previous one
     */
    private String CreateDocId(Date date)
    {
        long microseconds = scheduler.now() * 1000;
        lastMicroseconds  = (microseconds > lastMicroseconds) ? microseconds : lastMicroseconds + 1;
        return String.format("%1$014x%2$s", lastMicroseconds, suffix);
    }
 
    private String JsonNodeToString(JsonNode json)
    {
    	if (json == null) return null;
    	
    	JsonNodeType jsonType = json.getNodeType();
    	if (jsonType == JsonNodeType.STRING) return json.asText();
    	return (String) ((POJONode) json).getPojo();
    }
}
